### PR TITLE
PWG-UD: Single Gap producer

### PR DESCRIPTION
**Pull request by abylinkin**
New Table producer written to study inclusive nuclear reactions based on requiring FIT veto on one side of the event only.

### CODE REVIEW

**Review by aalkin — APPROVED** [review]
Approving to start the tests, but the merging decision is up to PWGUD representative.

File: PWGUD/DataModel/UDTables.h [hunk 1]
```diff
@@ -149,6 +151,18 @@ DECLARE_SOA_TABLE(UDCollisions, "AOD", "UDCOLLISION",
                   udcollision::NetCharge,
                   udcollision::RgtrwTOF);
 
+DECLARE_SOA_TABLE(SGCollisions, "AOD", "SGCOLLISION",
```
Comment (pbuehler) [inline PWGUD/DataModel/UDTables.h:154]: UDCOllision and SGCollision are equal except for the column GapSide. I propose that you remove table SGCollision but add a table e.g. CillisionType which can be joined with UDCollisions and contains one column with the GapSide information.

File: Tutorials/PWGMM/myExampleTask.cxx
Comment (pbuehler) [inline Tutorials/PWGMM/myExampleTask.cxx:1]: I think you should include this example task in the PR. Add only changes which shall become part of O2Physics.

File: Tutorials/PWGUD/CMakeLists.txt
Comment (pbuehler) [inline Tutorials/PWGUD/CMakeLists.txt:1]: Same comment as above. Remove this from the PR.

File: Tutorials/PWGUD/UDTutorial_01sg.cxx
Comment (pbuehler) [inline Tutorials/PWGUD/UDTutorial_01sg.cxx:1]: Remove this from PR.

File: Tutorials/PWGUD/UDTutorial_03b.cxx
Comment (pbuehler) [inline Tutorials/PWGUD/UDTutorial_03b.cxx:1]: Remove this from PR.

**Review by pbuehler — CHANGES_REQUESTED** [review]
Hi Sasha,
please have a look at my suggestions and do not hesitate to contact me if you have questions.

File: PWGUD/DataModel/UDTables.h (hunk 1 above)
Comment (abylinkin) [reply PWGUD/DataModel/UDTables.h:154]: Done

File: PWGUD/Core/SGSelector.h [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+
+  // Function to check if BC passes SG filter (without associated collision)
+  template <typename BCs, typename TCs, typename FWs>
+  int IsSelected(SGCutParHolder diffCuts, BCs& bcRange, TCs& tracks, FWs& fwdtracks, bool useSideA)
+  {
+    // check that there are no FIT signals in bcRange
+    // Single Gap (SG) condition
+    for (auto const& bc : bcRange) {
+      LOGF(info, "%i \t %i", udhelpers::cleanFITA(bc, diffCuts.maxFITtime(), diffCuts.FITAmpLimits()), udhelpers::cleanFITC(bc, diffCuts.maxFITtime(), diffCuts.FITAmpLimits()));
+      if (!udhelpers::cleanFITA(bc, diffCuts.maxFITtime(), diffCuts.FITAmpLimits()) && !udhelpers::cleanFITC(bc, diffCuts.maxFITtime(), diffCuts.FITAmpLimits())) {
```
[the remark this answers is not in the document]
Comment (abylinkin) [reply PWGUD/Core/SGSelector.h]: Part of the code rewritten

File: PWGUD/Core/SGSelector.h [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+    // net charge
+    auto netChargeValues = diffCuts.netCharges();
+    if (std::find(netChargeValues.begin(), netChargeValues.end(), netCharge) == netChargeValues.end()) {
+      return 11;
+    }
+    // invariant mass
+    if (ivm.M() < diffCuts.minIVM() || ivm.M() > diffCuts.maxIVM()) {
+      return 12;
+    }
+    // check that there are no FIT signals in any of the compatible BCs
```
[the remark this answers is not in the document]
Comment (abylinkin) [reply PWGUD/Core/SGSelector.h]: Part of the code rewritten in another manner

File: PWGUD/Core/SGSelector.h [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+  template <typename CC, typename BCs, typename TCs, typename FWs>
+  int IsSelected(SGCutParHolder diffCuts, CC& collision, BCs& bcRange, TCs& tracks, FWs& fwdtracks, bool useSideA)
+  {
+    LOGF(debug, "Collision %f", collision.collisionTime());
+    LOGF(debug, "Number of close BCs: %i", bcRange.size());
+    // check that there are no FIT signals in any of the compatible BCs
+    // Single Gap (SG) condition
+    int sg = 1;
+    int dg = 1;
+    for (auto const& bc : bcRange) {
```
[the remark this answers is not in the document]
Comment (abylinkin) [reply PWGUD/Core/SGSelector.h:47]: Done

**Review by pbuehler — APPROVED** [review]

**Pull request merged**